Make CUDAFuture remember and restore current device in callback

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #49014 Drop FutureNCCL in favor of vanilla CUDAFuture
* **#48789 Make CUDAFuture remember and restore current device in callback**
* #48840 Remove DataPtr extractor from CUDAFuture
* #48788 Cache the DataPtrs in CUDAFuture
* #48506 Split out reusable CUDAFuture from FutureNCCL
* #48505 Merge common parts of FutureNCCL into at::ivalue::Future
* #48504 Split FutureNCCL's CUDA-specific parts from generic future logic
* #48502 Support wider range of types in FutureNCCL
* #48501 Don't store device indices separately on FutureNCCL
* #48500 Add multi-GPU support to FutureNCCL
* #48563 Fix FutureNCCL not recording dataptrs with caching alloc in wait()
* #48503 Fix FutureNCCL's completed() disagreeing with wait()
* #48499 Record CUDA events for "follow-up" FutureNCCL inside markCompleted
* #48498 Use fresh stream from pool for each FutureNCCL callback
* #48497 Make FutureNCCL record events in current stream
* #48496 Have FutureNCCL record streams w/ allocator in addCallback
* #48562 Add some safeguards to FutureNCCL
* #48495 Remove NCCL dependency from PythonFutureWrapper
* #48561 Avoid using FutureNCCL before it's ready

CUDAFuture aims to "capture" the current state of CUDA-related stuff when the future is marked complete (e.g., by looking at current streams and recording events on them) and then "replicate" a similar state when users synchronize with the result of the future (by synchronizing the current streams with these events).

However, one "contextual" aspect of CUDA that we weren't capturing/replicating was the current device. This diff tries to fix that. I must mention that we can only do this for callbacks, while we cannot do it for the wait() method. I don't know if such a discrepancy between the two actually makes the overall behavior _worse_. I'd love to hear people's opinions on this.

Differential Revision: [D25210335](https://our.internmc.facebook.com/intern/diff/D25210335/)

lw committed:
Make CUDAFuture remember and restore current device in callback

CUDAFuture aims to "capture" the current state of CUDA-related stuff when the future is marked complete (e.g., by looking at current streams and recording events on them) and then "replicate" a similar state when users synchronize with the result of the future (by synchronizing the current streams with these events).

However, one "contextual" aspect of CUDA that we weren't capturing/replicating was the current device. This diff tries to fix that. I must mention that we can only do this for callbacks, while we cannot do it for the wait() method. I don't know if such a discrepancy between the two actually makes the overall behavior _worse_. I'd love to hear people's opinions on this.

Differential Revision: [D25210335](https://our.internmc.facebook.com/intern/diff/D25210335/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDAFuture remember and restore current device in callback"


CUDAFuture aims to "capture" the current state of CUDA-related stuff when the future is marked complete (e.g., by looking at current streams and recording events on them) and then "replicate" a similar state when users synchronize with the result of the future (by synchronizing the current streams with these events).

However, one "contextual" aspect of CUDA that we weren't capturing/replicating was the current device. This diff tries to fix that. I must mention that we can only do this for callbacks, while we cannot do it for the wait() method. I don't know if such a discrepancy between the two actually makes the overall behavior _worse_. I'd love to hear people's opinions on this.

Differential Revision: [D25210335](https://our.internmc.facebook.com/intern/diff/D25210335/)

[ghstack-poisoned]

lw committed:
Update on "Make CUDAFuture remember and restore current device in callback"


CUDAFuture aims to "capture" the current state of CUDA-related stuff when the future is marked complete (e.g., by looking at current streams and recording events on them) and then "replicate" a similar state when users synchronize with the result of the future (by synchronizing the current streams with these events).

However, one "contextual" aspect of CUDA that we weren't capturing/replicating was the current device. This diff tries to fix that. I must mention that we can only do this for callbacks, while we cannot do it for the wait() method. I don't know if such a discrepancy between the two actually makes the overall behavior _worse_. I'd love to hear people's opinions on this.

Differential Revision: [D25210335](https://our.internmc.facebook.com/intern/diff/D25210335/)

[ghstack-poisoned]

mrshenli commented on aten/src/ATen/cuda/CUDAFuture.h:
```diff
@@ -31,6 +31,8 @@ struct TORCH_CUDA_API CUDAFuture : at::ivalue::Future {
   }
 
   void postMarkCompletedHook(const at::IValue& value) override {
+    currentDevice_ = c10::cuda::current_device();
```
hmm, why we are recording the device when the future is marked as completed, instead of remember the device when the callback was inserted (through `then`/`addCallback`)?

lw replied:
I guess those are two different approaches to this. The "philosophy" I was following is this: `then()` and `addCallback()` are used to perform a computation after another one is complete. If we were dealing with with sync operations, one would do this:
```
do_sth_sync()
do_sth_later()
```
but with async ops this needs to change and become
```
fut = do_sth_async()
fut.then(do_sth_later)
```
In the sync scenario, `do_sth_later()` runs in the same "environment"/"context" as `do_sth_sync()` (same current device, same current streams, ...). In this diff I was trying to recreate this in the async case, by "recording" the environment's state at the end of the async operation, and then "recreating" it in the callback.

Note that the approach we have for streams is _somewhat_ similar. We do _not_ run the callback in the streams that were current when the callback was inserted. (In honesty, we also do not run it in the streams that were current when the async op finished (as that would mean running computations in the I/O streams), but we do synchronize the "fresh" streams with those streams).

Note also that I am not opposed to changing this behavior, and replicate the "environment" that was current when the user inserted the callback. That has its own sets of advantages. For streams, for example, it allows the user to very precisely control what streams the callback will use.

mrshenli replied:
> For streams, for example, it allows the user to very precisely control what streams the callback will use.

Yep, same for device as well, especially when the callback fn is an imported function and users cannot easily change it. The current behavior should be sufficient to unblock RPC use cases. So I am OK to land this PR and modify it later if necessary.

lw committed:
Update on "Make CUDAFuture remember and restore current device in callback"


CUDAFuture aims to "capture" the current state of CUDA-related stuff when the future is marked complete (e.g., by looking at current streams and recording events on them) and then "replicate" a similar state when users synchronize with the result of the future (by synchronizing the current streams with these events).

However, one "contextual" aspect of CUDA that we weren't capturing/replicating was the current device. This diff tries to fix that. I must mention that we can only do this for callbacks, while we cannot do it for the wait() method. I don't know if such a discrepancy between the two actually makes the overall behavior _worse_. I'd love to hear people's opinions on this.

Differential Revision: [D25210335](https://our.internmc.facebook.com/intern/diff/D25210335/)

[ghstack-poisoned]
